### DISCUSSION TITLE
Docker development with the database in a RAM disk is a huge speedup

### DISCUSSION THREAD

**Discussion by trombonekenny** [PrairieLearn/PrairieLearn#7078] [Dev]
FYI: I made https://github.com/PrairieLearn/PrairieLearn/pull/5751 to help the existing in-docker scripts work better with a postgres database in a non-standard location. The PR describes how I make a RAM disk (tmpfs volume in docker) to hold the database. I/O is really what kills docker containers, and it's not a good practice to put a database inside a docker volume (but we do it for convenience). It's better to mount it externally so it doesn't have the layer writing overhead. Doing it to a RAM disk is even nicer performance-wise.

With the DB pointing to /var/postgres inside the container, a full `make test` run takes me 5 minutes 38 seconds.
If I point the database to a RAM disk, that drops to 1 minute 43 seconds.

**Comment by echuber2** [comment]
Something similar can be done for `node_modules` too, although the `--tmpfs` option on Docker seems to create a mount with `noexec` by default which causes problems. Manually creating a directory ahead of time under `/dev/shm` and using it for the mount seems to work, if `/dev/shm` doesn't also have `noexec`, or a `tmpfs` could be created separately.

**Reply by trombonekenny** [reply]
In a dev context, `node_modules` comes in with my `-v $PWD:/PrairieLearn` so it's already better than the layered in-docker disk. I didn't time compare mounting in an external volume versus an external tmpfs volume but both are going to better than IO inside the container.

**Reply by trombonekenny** [reply]
Nathan and I discovered that pointing --tmpfs at an existing path in the container gave it different starting permissions than a new path. (From Ubuntu 18.04 LTS)

```
✔ ~/git/PrairieLearn [lti13draft1 L|✚ 2⚑ 3]
15:30 discovery$ docker run -it --rm --tmpfs=/var/postgrest -v $PWD:/PrairieLearn prairielearn/prairielearn /bin/bash
bash-4.2# cd /var
bash-4.2# ls -ld post*
drwx------ 1 postgres postgres 4096 Feb  2 14:17 postgres
drwxrwxrwt 2 root     root       40 Feb  3 21:30 postgrest
bash-4.2# exit
exit
✔ ~/git/PrairieLearn [lti13draft1 L|✚ 2⚑ 3]
15:31 discovery$ docker run -it --rm --tmpfs=/var/postgres -v $PWD:/PrairieLearn prairielearn/prairielearn /bin/bash
bash-4.2# cd /var
bash-4.2# ls -ld post*
drwx------ 2 root root 40 Feb  3 21:31 postgres
bash-4.2# 
```

We made some last minute changes to `start_postgres.sh` so that it chown/chmod's the directory first, and then `--tmpfs=/var/postgres` worked. We'd probably run into that same problem with `node_modules` unless you could use an environment variable or config to point it somewhere else?

**Comment by trombonekenny** [comment]
The `docker/start_postgres.sh` changes have landed in `master`. You can call docker with `--tmpfs=/var/postgres` or `--tmpfs=/pgtmpfs -e PGDATA=/pgtmpfs` and the script will "do the right thing" to initialize the database when it starts up. Try it and see if it helps in your environment!
